[no-undef-components] false positive for "setup" components

**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 8.23.1
- **eslint-plugin-vue version:** 9.5.0
- **Node version:** 18.4.0
- **Operating System:** Mac OS

**Please show your full configuration:**
<!-- Paste content of your .eslintrc file -->
```json5
{
    "parserOptions": {
        "parser": "@typescript-eslint/parser",
        "extraFileExtensions": [".vue"]
    },

    "extends": ["plugin:vue/base"],

    "rules": {
        "vue/no-undef-components": "error"
    }
}
```

**What did you do?**	
<!-- Please include the actual source code causing the issue. -->
```vue
<template>
  <AppHeader />
</template>

<script setup lang="ts">
  import AppHeader from '@/components/AppHeader.vue';
</script>
```

**What did you expect to happen?**

No error produced.

**What actually happened?**

```
/private/tmp/eslint-vue-no-undef-components-bug/App.vue
  2:3  error  The '<AppHeader>' component has been used, but 'AppHeader' only refers to a type  vue/no-undef-components

✖ 1 problem (1 error, 0 warnings)
```

**Repository to reproduce this issue**

https://github.com/Ky6uk/eslint-vue-no-undef-components-bug

--- a/lib/rules/no-undef-components.js
+++ b/lib/rules/no-undef-components.js
@@ -133,7 +133,29 @@ module.exports = {
           (scope) => scope.type === 'module'
         )
         for (const variable of (moduleScope && moduleScope.variables) || []) {
-          if (variable.isTypeVariable) {
+          if (
+            // Check for type definitions. e.g. type Foo = {}
+            (variable.isTypeVariable && !variable.isValueVariable) ||
+            // type-only import seems to have isValueVariable set to true. So we need to check the actual Node.
+            (variable.defs.length > 0 &&
+              variable.defs.every((def) => {
+                if (def.type !== 'ImportBinding') {
+                  return false
+                }
+                if (def.parent.importKind === 'type') {
+                  // check for `import type Foo from './xxx'`
+                  return true
+                }
+                if (
+                  def.node.type === 'ImportSpecifier' &&
+                  def.node.importKind === 'type'
+                ) {
+                  // check for `import { type Foo } from './xxx'`
+                  return true
+                }
+                return false
+              }))
+          ) {
             scriptTypeOnlyNames.add(variable.name)
           } else {
             scriptVariableNames.add(variable.name)

--- a/tests/lib/rules/no-undef-components.js
+++ b/tests/lib/rules/no-undef-components.js
@@ -6,6 +6,7 @@
 
 const RuleTester = require('eslint').RuleTester
 const rule = require('../../../lib/rules/no-undef-components')
+const semver = require('semver')
 
 const tester = new RuleTester({
   parser: require.resolve('vue-eslint-parser'),
@@ -681,9 +682,14 @@ tester.run('no-undef-components', rule, {
         }
       ]
     },
-    {
-      filename: 'test.vue',
-      code: `
+    ...(semver.gte(
+      require('@typescript-eslint/parser/package.json').version,
+      '5.0.0'
+    )
+      ? [
+          {
+            filename: 'test.vue',
+            code: `
       <script setup lang="ts">
         import type Foo from './Foo.vue'
         import type {HelloWorld1} from './components/HelloWorld'
@@ -702,45 +708,47 @@ tester.run('no-undef-components', rule, {
         <HelloWorld5 />
       </template>
       `,
-      parserOptions: {
-        ecmaVersion: 6,
-        sourceType: 'module',
-        parser: require.resolve('@typescript-eslint/parser')
-      },
-      parser: require.resolve('vue-eslint-parser'),
-      errors: [
-        {
-          message:
-            "The '<Foo>' component has been used, but 'Foo' only refers to a type.",
-          line: 12
-        },
-        {
-          message:
-            "The '<HelloWorld1>' component has been used, but 'HelloWorld1' only refers to a type.",
-          line: 13
-        },
-        {
-          message:
-            "The '<HelloWorld2>' component has been used, but 'HelloWorld2' only refers to a type.",
-          line: 14
-        },
-        {
-          message:
-            "The '<HelloWorld3>' component has been used, but 'HelloWorld3' only refers to a type.",
-          line: 15
-        },
-        {
-          message:
-            "The '<HelloWorld4>' component has been used, but 'HelloWorld4' only refers to a type.",
-          line: 16
-        },
-        {
-          message:
-            "The '<HelloWorld5>' component has been used, but 'HelloWorld5' only refers to a type.",
-          line: 17
-        }
-      ]
-    },
+            parserOptions: {
+              ecmaVersion: 6,
+              sourceType: 'module',
+              parser: require.resolve('@typescript-eslint/parser')
+            },
+            parser: require.resolve('vue-eslint-parser'),
+            errors: [
+              {
+                message:
+                  "The '<Foo>' component has been used, but 'Foo' only refers to a type.",
+                line: 12
+              },
+              {
+                message:
+                  "The '<HelloWorld1>' component has been used, but 'HelloWorld1' only refers to a type.",
+                line: 13
+              },
+              {
+                message:
+                  "The '<HelloWorld2>' component has been used, but 'HelloWorld2' only refers to a type.",
+                line: 14
+              },
+              {
+                message:
+                  "The '<HelloWorld3>' component has been used, but 'HelloWorld3' only refers to a type.",
+                line: 15
+              },
+              {
+                message:
+                  "The '<HelloWorld4>' component has been used, but 'HelloWorld4' only refers to a type.",
+                line: 16
+              },
+              {
+                message:
+                  "The '<HelloWorld5>' component has been used, but 'HelloWorld5' only refers to a type.",
+                line: 17
+              }
+            ]
+          }
+        ]
+      : []),
     {
       filename: 'test.vue',
       code: `

--- a/tests/lib/rules/no-undef-components.js
+++ b/tests/lib/rules/no-undef-components.js
@@ -598,6 +598,26 @@ tester.run('no-undef-components', rule, {
       <script setup>
       import TheModal from 'foo'
       </script>`
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script setup lang="ts">
+        import Foo from './Foo.vue'
+        import {HelloWorld1} from './components/HelloWorld'
+      </script>
+
+      <template>
+        <Foo />
+        <HelloWorld1 />
+      </template>
+      `,
+      parserOptions: {
+        ecmaVersion: 6,
+        sourceType: 'module',
+        parser: require.resolve('@typescript-eslint/parser')
+      },
+      parser: require.resolve('vue-eslint-parser')
     }
   ],
   invalid: [
@@ -718,6 +738,31 @@ tester.run('no-undef-components', rule, {
           message:
             "The '<HelloWorld5>' component has been used, but 'HelloWorld5' only refers to a type.",
           line: 17
+        }
+      ]
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <script setup lang="ts">
+        type Foo = {}
+      </script>
+
+      <template>
+        <Foo />
+      </template>
+      `,
+      parserOptions: {
+        ecmaVersion: 6,
+        sourceType: 'module',
+        parser: require.resolve('@typescript-eslint/parser')
+      },
+      parser: require.resolve('vue-eslint-parser'),
+      errors: [
+        {
+          message:
+            "The '<Foo>' component has been used, but 'Foo' only refers to a type.",
+          line: 7
         }
       ]
     },

--- a/typings/eslint-plugin-vue/util-types/ast/es-ast.ts
+++ b/typings/eslint-plugin-vue/util-types/ast/es-ast.ts
@@ -274,6 +274,7 @@ export interface ImportSpecifier extends HasParentNode {
   type: 'ImportSpecifier'
   imported: Identifier
   local: Identifier
+  importKind?: 'type' | 'value'
 }
 export interface ImportDefaultSpecifier extends HasParentNode {
   type: 'ImportDefaultSpecifier'

